### PR TITLE
feat(core): add CDI policy resolver

### PR DESCRIPTION
**Pull request by elezar**
## Summary

Add a portable, fail-closed CDI context resolver that turns validated CDI specs into policy-relevant device, mount, and group requirements. This is the foundation of the stacked #1606 implementation.

## Related Issue

Part of #1606.

## Changes

- Add the shared CDI context schema and Linux-only resolver dependency boundary.
- Validate device-node types, paths, writable mounts, and supplemental GIDs.
- Add resolver unit coverage using temporary CDI specifications.

## Testing

- [x] `mise run pre-commit` passes
- [x] Unit tests added/updated
- [ ] E2E tests added/updated (not applicable to this layer)

## Checklist

- [x] Follows Conventional Commits
- [x] Commits are signed off (DCO)

### CODE REVIEW

File: crates/openshell-core/src/cdi_linux.rs [hunk 1]
```diff
@@ -0,0 +1,846 @@
...
+        }
+    }
+    parsed
+}
+
+fn resolve_container_edits(
+    context: &CdiContext,
+    selected_devices: &[String],
+) -> Result<CdiContainerEdits, CdiError> {
+    let (mut cache, refresh_error) = build_cache(&context.spec_dirs);
```
Comment (drew) [inline crates/openshell-core/src/cdi_linux.rs:237]: [P1] Fail on cache refresh errors before resolving selected devices. The pinned CDI cache can return an error while retaining a resolvable device (for example, two equal-priority specs defining `nvidia.com/gpu=0`). Because `refresh_error` is only consulted when lookup fails, this resolver accepts one ambiguous spec instead of failing closed; I reproduced this with two conflicting YAML specs and `resolve_cdi_context` returned `Ok`. Reject the refresh error here, or prove specifically that the selected device was unaffected, and add a duplicate-selected-device regression test.

File: crates/openshell-core/src/cdi_linux.rs [hunk 2]
```diff
@@ -0,0 +1,846 @@
...
+            Some(_) => Ok(()),
+            None => {
+                self.mount_paths.insert(path, access);
+                Ok(())
+            }
+        }
+    }
+
+    fn add_gid(&mut self, gid: u32) {
+        if gid == 0 {
```
Comment (drew) [inline crates/openshell-core/src/cdi_linux.rs:120]: [P2] Reject root supplemental GID instead of silently dropping it. The accepted #1606 implementation plan requires GID 0 to fail validation. Skipping it makes an unsafe CDI request look valid and prevents the supervisor from emitting its validation finding. Return a dedicated `CdiError` and change the existing root-GID test to expect rejection.

**Review by drew — CHANGES_REQUESTED** [review]
Two fail-closed validation issues need correction before this layer merges. I reproduced the cache-refresh case with a focused openshell-core test: two conflicting specs defining the selected device currently return Ok.

File: crates/openshell-core/src/cdi_linux.rs (hunk 1 above)
Comment (elezar) [reply crates/openshell-core/src/cdi_linux.rs:237]: The issue here is that the building the cache loads all CDI specs from the specified spec directories. These could include specs from different vendors or devices. Furthermore, if a spec fails to load (e.g due to malfomed YAML) there is no way to know which vendor or devices it was intended for.

This is the reason that we retain this information and include it if a device is unresolvable. This behaviour aligns with the behaviour for other CDI consumers including Docker, Podman, Containerd, Crio, and Kata.

I can add a test to prove that duplicate devices of the same priority causes a failure though.

File: crates/openshell-core/src/cdi_linux.rs (hunk 2 above)
Comment (elezar) [reply crates/openshell-core/src/cdi_linux.rs:120]: That's a fair point. When generating a CDI spec, the NVIDIA Contianer Toolkit [skips `gid==0`](https://github.com/NVIDIA/nvidia-container-toolkit/blob/5eee4c5f30d8448f0c968402b2ea789ef9a0e20f/internal/edits/device.go#L127) but there is no guarantee that other vendors do.

Note that the the Go implementation of the [CDI injection code](https://github.com/cncf-tags/container-device-interface/blob/4b4ceb27b3b9a1baf9b0c60d82ec588cff122750/pkg/cdi/container-edits.go#L177) also skips `gid==0`.

**Review by drew — DISMISSED** [review]

**Comment by cjagwani** [comment]
The current required E2E run (34984111251) is red in three jobs that appear unrelated to the CDI resolver: the WSL GPU sandbox create timed out, rust-docker timed out in the download-only case, and oidc-pkce-docker hit SQLite `database is locked`. Branch checks and the Linux ARM/AMD GPU jobs passed. I attempted `gh run rerun 34984111251 --failed`, but this account does not have the repository admin permission GitHub requires. Could an OpenShell maintainer rerun the failed jobs so we can distinguish flakes from a current-head regression?

File: crates/openshell-core/src/cdi_linux.rs [hunk 3]
```diff
@@ -0,0 +1,977 @@
...
+        Ok(())
+    }
+
+    fn build(self) -> CdiDerivedRequirements {
+        let mut read_only_paths = BTreeSet::new();
+        let mut read_write_mount_paths = Vec::new();
+        for (path, access) in self.mount_paths {
+            match access {
+                CdiAccess::ReadOnly => {
+                    read_only_paths.insert(shared_library_parent(&path).unwrap_or(path));
```
Comment (drew) [inline crates/openshell-core/src/cdi_linux.rs:123]: > **gator-agent**

**Warning — GATOR-ed094c94-01 · Revalidate paths after library-parent expansion**

**Summary:** A selected CDI spec can mount a shared library directly below a prohibited broad directory, such as `/libcuda.so.1`. The resolver validates the file path, then replaces it with `/` without reapplying the broad-path guard. The stacked consumer would grant read access to the entire workload filesystem; `/usr/libcuda.so.1` similarly recreates the prohibited `/usr` grant.

**Fix:** Validate the expanded parent with the same unsafe-path rules before inserting it, and reject prohibited parents. Add regressions for `/libcuda.so.1` and `/usr/libcuda.so.1`.

**Verify:** Resolve a selected spec containing each mount. Current head returns `/` or `/usr` in `read_only_paths`; the fixed resolver must return `UnsafePolicyPath`.

<details>
<summary>Agent context</summary>

- **Ownership:** This PR performs parent expansion after its only unsafe-path validation point.
- **Location:** `crates/openshell-core/src/cdi_linux.rs:123`

</details>

**Review by drew — COMMENTED** [review]
> **gator-agent**

## PR Review Status

This roadmap-backed core layer is project-valid, and I reviewed it with #2776 as the downstream stack consumer. Thanks @elezar—I checked your explanation that cache refresh errors can originate in unrelated vendor specs; because #1606 makes malformed or ambiguous CDI input fail closed and the consumer treats resolver success as permission to launch, the remaining refresh behavior still needs correction.

Action required: please address the two fail-closed resolver findings below and push an updated head.

Blocking findings:
- `GATOR-ed094c94-01`: shared-library parent expansion can recreate a prohibited broad policy path.
- `GATOR-ed094c94-02`: selected devices can resolve successfully without a complete, successful cache refresh.

Carried findings:
- None

Non-blocking suggestions:
- None

<details>
<summary>Gator metadata</summary>

- Validation: Project-valid as the maintainer-authored shared-core layer for roadmap issue #1606 and stacked consumer PR #2776.
- Docs: Not needed in this internal resolver layer; the user-visible consumer documentation is in #2776.
- Checks: Current-head Branch Checks, Helm Lint, and Trivy Changes are green.
- E2E: `test:e2e` and `test:e2e-gpu` are applied; both current-head required gates are green.
- Head SHA: `ed094c94e95b75058836e3ba5c125a89e938cc14`
- Base SHA: `473d1e997439c3577fabd9014ba8e8959e3c2444`
- Merge base SHA: `473d1e997439c3577fabd9014ba8e8959e3c2444`
- Patch ID: `9e2add7e653940199201e01aab9880af26a1a42f`
- Gator payload: `10`
- Review mode: `initial`
- Previous reviewed SHA: none
- Review budget exhausted: no
- Maintainer decision required: no
- Next state: `gator:in-review`

</details>

File: crates/openshell-core/src/cdi_linux.rs [hunk 4]
```diff
@@ -0,0 +1,977 @@
...
+            read_write_mount_paths,
+            additional_gids: self.additional_gids.into_iter().collect(),
+        }
+    }
+}
+
+pub fn resolve_cdi_context(context: &CdiContext) -> Result<CdiDerivedRequirements, CdiError> {
+    validate_context(context)?;
+    let selected_devices = selected_cdi_devices(&context.selected_devices);
+    if selected_devices.is_empty() || context.spec_dirs.is_empty() {
```
Comment (drew) [inline crates/openshell-core/src/cdi_linux.rs:140]: > **gator-agent**

**Warning — GATOR-ed094c94-02 · Require complete CDI discovery before success**

**Summary:** A compute driver can request a CDI device while providing no spec directories, and the resolver reports empty success. Separately, a malformed or conflicting projected spec can make refresh fail while a retained selected device still resolves. The downstream sandbox then launches with missing, incomplete, or ambiguous device policy instead of failing closed.

**Fix:** Error when selection is non-empty but spec directories are empty. Also terminate on every cache refresh error before lookup, preserving the upstream diagnostic; retain missing-ID, malformed-spec, and duplicate-selected-device coverage.

**Verify:** Resolve a selected ID with no directories, then with a valid selected spec beside malformed YAML. Both currently can return `Ok`; both must deterministically fail.

<details>
<summary>Agent context</summary>

- **Ownership:** This PR defines the new resolver success boundary and suppresses these incomplete-discovery states.
- **Location:** `crates/openshell-core/src/cdi_linux.rs:140`

</details>

**Review by drew — COMMENTED** [review]
> **gator-agent**

## PR Review Status

I rechecked the rebased head in stack context with #2776. The author-only delta updates lockfile dependency versions but does not change either fail-closed CDI resolver path, so both existing findings remain open; no new findings were added.

Action required: @elezar, please address the two carried findings in their existing review threads and push an updated head.

Blocking findings:
- No new blocking findings.

Carried findings:
- `GATOR-ed094c94-01`: shared-library parent expansion still allows a prohibited broad policy path.
- `GATOR-ed094c94-02`: non-empty device selection can still succeed without complete CDI discovery and a successful cache refresh.

<details>
<summary>Gator metadata</summary>

- Validation: Project-valid as the maintainer-authored shared-core layer for roadmap issue #1606 and stacked consumer PR #2776.
- Docs: Not needed in this internal resolver layer; the user-visible consumer documentation is in #2776.
- Checks: Current-head Branch Checks, Helm Lint, and Trivy Changes are green.
- E2E: `test:e2e` and `test:e2e-gpu` are applied; both current-head required gates are green.
- Head SHA: `eab54064acd753a8cfda0f92e28bbdb4d0994e00`
- Base SHA: `d02ebe2c4b22068960acefc28a2515be97d24c17`
- Merge base SHA: `d02ebe2c4b22068960acefc28a2515be97d24c17`
- Patch ID: `a3c8870dc3ad0c632af919ce155841e065bfc3b6`
- Gator payload: `10`
- Review mode: `follow_up`
- Previous reviewed SHA: `ed094c94e95b75058836e3ba5c125a89e938cc14`
- Review budget exhausted: no
- Maintainer decision required: no
- Next state: `gator:in-review`

</details>

**Review by drew — COMMENTED** [review]
> **gator-agent**

## PR Review Status

The current head is a rebase-equivalent copy of the already reviewed CDI resolver patch, so no new code-review round was needed. Evaluating it with stacked consumer PR #2776 confirms that both existing fail-closed obligations remain relevant: the consumer still turns resolver output into launch policy, while the resolver patch has not changed either affected path.

Action required: @elezar, please address the two carried findings in their existing review threads and push an updated head.

Blocking findings:
- No new blocking findings.

Carried findings:
- `GATOR-ed094c94-01`: shared-library parent expansion still allows a prohibited broad policy path.
- `GATOR-ed094c94-02`: non-empty device selection can still succeed without complete CDI discovery and a successful cache refresh.

<details>
<summary>Gator metadata</summary>

- Validation: Project-valid as the maintainer-authored shared-core layer for roadmap issue #1606 and stacked consumer PR #2776.
- Docs: Not needed in this internal resolver layer; the user-visible consumer documentation is in #2776.
- Checks: Branch Checks, Helm Lint, Trivy Changes, and GPU E2E are green; Core E2E is failing in the rust-docker test job. Review feedback remains the active gate.
- E2E: `test:e2e` and `test:e2e-gpu` are applied; the current-head GPU gate is green and the core gate is red.
- Head SHA: `ff1ffe791bce60d38a29c429925231fb435ed31c`
- Base SHA: `069ae6bd966c4ced02b5cff0a26aee0a7f23a82f`
- Merge base SHA: `069ae6bd966c4ced02b5cff0a26aee0a7f23a82f`
- Patch ID: `a3c8870dc3ad0c632af919ce155841e065bfc3b6`
- Gator payload: `10`
- Review mode: `already_reviewed` (rebase-equivalent)
- Previous reviewed SHA: `eab54064acd753a8cfda0f92e28bbdb4d0994e00`
- Review budget exhausted: no
- Maintainer decision required: no
- Next state: `gator:in-review`
- Blocked reason: `author_changes_required`

</details>
